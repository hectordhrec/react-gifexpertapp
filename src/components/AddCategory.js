import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        //setInputValue(e.target.value);
        setInputValue(valor => e.target.value);
        console.log('handleInputchange');
    }

    const handleSunmit = (e) => {
        e.preventDefault();
        if (inputValue.trim().length > 2) {
            setCategories(categories => [inputValue,...categories]);
            setInputValue('');
        }
    }

    return (
        <form onSubmit={ handleSunmit }>
            <p>{inputValue}</p>
            <input type="text" className="form-control" value={inputValue} onChange={handleInputChange} />
        </form>
    );
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}

