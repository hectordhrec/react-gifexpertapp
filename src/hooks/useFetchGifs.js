import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    /** Al colocar la categorÃ­a como dependencia, cada vez que la categorÃ­a cambie se va a ejecutar el useEffect */
    useEffect(() => {
        getGifs(category).then(imgs => {
            setState({
                data: imgs,
                loading: false
            });
        });
    }, [category]);

    return state;
}