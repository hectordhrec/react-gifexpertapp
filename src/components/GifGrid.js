import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
import PropTypes from 'prop-types';

export const GifGrid = ({ category }) => {
    // const [images, setImages] = useState([]);

    /** Al colocar la categoría como dependencia, cada vez que la categoría cambie se va a ejecutar el useEffect */
    // useEffect(() => {
    //     getGifs(category).then(setImages);
    //     // getGifs().then(imgs => setImages(imgs)); // Otra forma
    // }, [category]);

    const { data: images, loading } = useFetchGifs(category);


    return (
        <>
            <h3 className="card animate__animated animate__fadeIn">{category}</h3>
            { loading && <p className="animate__animated animate__flash">Loading</p>}
            <div className="card-grid">
                {images.map((img) => (
                    <GifGridItem
                        key={img.id}
                        {...img}
                    />
                ))}
            </div>
        </>
    );
};

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}
