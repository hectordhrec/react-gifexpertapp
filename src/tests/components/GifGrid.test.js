import { shallow } from 'enzyme'
import { GifGrid } from '../../components/GifGrid'
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Prueba al componente <GifGrid />', () => {

    const category = 'bojack';
    
    test('Debe de mostrarse correctamente', () => {

        useFetchGifs.mockReturnValue({
            data: [],
            loading: false
        });

        const wrapper = shallow(<GifGrid category = { category } />);
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de mostrar items cuando se cargan imagenes useFetchGifs', () => {

        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/imagen',
            title: 'titulo'
        }];
        
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
        const wrapper = shallow(<GifGrid category = { category } />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
    });
    
    
});
