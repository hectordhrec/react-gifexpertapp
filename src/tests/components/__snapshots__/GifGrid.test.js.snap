// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba al componente <GifGrid /> Debe de mostrar items cuando se cargan imagenes useFetchGifs 1`] = `
<Fragment>
  <h3
    className="card animate__animated animate__fadeIn"
  >
    bojack
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="titulo"
      url="https://localhost/imagen"
    />
  </div>
</Fragment>
`;

exports[`Prueba al componente <GifGrid /> Debe de mostrarse correctamente 1`] = `
<Fragment>
  <h3
    className="card animate__animated animate__fadeIn"
  >
    bojack
  </h3>
  <div
    className="card-grid"
  />
</Fragment>
`;
